sympy/polys/polytools.py

```diff
@@ -7003,7 +7123,17 @@ def real_roots(f, multiple=True, radicals=True):
     .. [1] https://en.wikipedia.org/wiki/Casus_irreducibilis
     """
     try:
-        F = Poly(f, greedy=False)
+        if isinstance(f, Poly):
+            if extension and not f.domain.is_AlgebraicField:
+                F = Poly(f.expr, extension=True)
+            else:
+                F = f
+        else:
+            if extension:
+                F = Poly(f, extension=True)
+            else:
+                F = Poly(f, greedy=False)
```

Comment:
There should be a simpler way of writing this. One problem with the above is that we call `Poly(f.expr, extension=True)` if we have a Poly with domain QQ which is unnecessary.

It seems like something is wrong if we can't do all of this in one call. But I can't think what the call should be...

Reply:
One reason I have all these checks is because if we just did something like:
```python
if extension:
            F = Poly(f, extension=True)
else:
            F = Poly(f, greedy=False)
```

the issue is that if specify extension=True if the Poly is already made, it won't change it. E.g.

```
In [3]: f = x - sqrt(2)

In [4]: f
Out[4]: x - sqrt(2)

In [5]: Poly(f, x)
Out[5]: Poly(x - sqrt(2), x, domain='EX')

In [6]: Poly(f, x, extension=True)
Out[6]: Poly(x - sqrt(2), x, domain='QQ<sqrt(2)>')

In [7]: Poly(Poly(f, x), extension=True)
Out[7]: Poly(x - sqrt(2), x, domain='EX')

In [8]: Poly(Poly(f, x).expr, extension=True)
Out[8]: Poly(x - sqrt(2), x, domain='QQ<sqrt(2)>')
```

But as the last output shows, if you use the expr then you can construct the extension. It seems important to me to handle the case that extension=True and the user passes a Poly. I'm not sure if there is a better way to construct the domain in this case.

Reply:
Maybe if the input is a Poly it is reasonable to just ignore the other options like extension=True.

Either way I was more complaining that there isn't a simple way to do what is being done here that does not require handling 4 separate cases.